riot-rs-threads: cortex_m: make `sched()` less hacky

kaspar030 opened the pull request:
This moves the previusly manual critical section into `sched()`, and uses a "u128" to pass our values to the assembly.
(this adds an instruction or two, but on nrf52840dk, this is barely measurable).

elenaf9 commented on src/riot-rs-threads/src/arch/cortex_m.rs:
```diff
@@ -216,53 +212,43 @@ unsafe extern "C" fn PendSV() {
-            if let Some(pid) = (&*THREADS.as_ptr(cs)).runqueue.get_next() {
-                next_pid = pid;
-                break;
+        if let Some(res) = critical_section::with(|cs| {
+            let threads = &mut *THREADS.as_ptr(cs);
```
Why not use `THREADS.with_mut(|mut threads| { ... } )` here?

kaspar030 replied:
Because `THREADS.with_mut()` checks for unique access within a CS, which we bypass here (if we are inside PendSV, there can be no other critical section active).

elenaf9 replied:
Doesn't `THREADS.with_mut()` do the exact thing that we are doing here as well? I.e. create a critical section, borrow the mutex and then execute the closure on it?
What am I missing?

kaspar030 replied:
Ah, I was missing something, yeah, that sounds good actually.

kaspar030 replied:
for some reason [that](https://gist.github.com/kaspar030/0a74d4867ae9d48cfe81d56516aa265c)'s 15% slower.

elenaf9 replied:
That's weird. The only difference that I see is using a raw pointer vs mutable reference to the data in the ref cell...

If that always impacts the performance this strongly, maybe we should directly change the `EnsureOnce::with_mut` implementation to use a raw pointer?

kaspar030 replied:
yeah. I suggest we get some measurement & testing in place before going with too many changes.

kaspar030 replied:
... and, postpone this?

elenaf9 replied:
> yeah. I suggest we get some measurement & testing in place before going with too many changes.

> ... and, postpone this?

Yes, sounds good.

elenaf9 commented on src/riot-rs-threads/src/arch/cortex_m.rs:
```diff
@@ -216,53 +212,43 @@ unsafe extern "C" fn PendSV() {
+                    if next_pid == current_pid {
+                        return Some(0);
+                    }
+                    //println!("current: {} next: {}", current_pid, next_pid);
```
```suggestion
```

elenaf9 commented on src/riot-rs-threads/src/arch/cortex_m.rs:
```diff
@@ -216,53 +212,43 @@ unsafe extern "C" fn PendSV() {
+                    threads.current_thread = Some(next_pid);
+                    threads.threads[current_pid as usize].data.as_ptr()
+                } else {
+                    core::ptr::null()
```
I know that this is old code, but could we add a check in the asm code if r2 is `0` and skip the store commands in that case?

kaspar030 replied:
that's already happening, if there's not "next_sp" (r0 == 0), the save/restore is skipped.

elenaf9 replied:
`next_sp` is 0 if the next thread equals the current thread.
I was talking about the case when `r2` is `0` because there was no current thread before.

But I just noticed that right now we never have the case that `current_thread` is `None` for cortex-m, because we set it already in `start_threading`, before the first interrupt.
But as part of #182 I am proposing to change this.

kaspar030 replied:
fixed in e3966d2864a57a49628b9742e31ac3a9ad8f218c.

elenaf9 commented on src/riot-rs-threads/src/arch/cortex_m.rs:
```diff
@@ -216,53 +212,43 @@ unsafe extern "C" fn PendSV() {
+                    (next_sp as u128) |  ((current_high_regs as u128) << 32) | ((next_high_regs as u128) << 64);
+                Some(res)
+            } else {
+                cortex_m::asm::wfi();
```
Shouldn't this be outside of the critical section (end of the loop, line 253), so that interrupts are enabled again?

kaspar030 replied:
It doesn't have to. "wfi" will always stall the CPU until an ISR is *pending*. This code here would then continue, end the critical section, which would then reenable ISRs and jump into them.

elenaf9 replied:
Ahh okay. I'd still argue that the code is easier to understand if the wfi happens below, but I don't feel strongly about this :).

kaspar030 replied:
That might actually cause a race. "critical section ends -> isr -> wfi", the wfi could block until the next isr comes in.

elenaf9 replied:
Ahh good point, I didn't think of that scenario.

elenaf9 commented on src/riot-rs-threads/src/arch/cortex_m.rs:
```diff
@@ -216,53 +212,42 @@ unsafe extern "C" fn PendSV() {
-                break;
+        if let Some(res) = critical_section::with(|cs| {
+            let threads = &mut *THREADS.as_ptr(cs);
+            if let Some(next_pid) = threads.runqueue.get_next() {
```
Nit: You could use match here and use an early return in case of `None` to make the code more readable.

kaspar030 replied:
how do you mean?

elenaf9 replied:
Like this: 
```rs
   let next_id = match threads.runqueue.get_next() {
        Some(pid) => pid,
        None => {
            cortex_m::asm::wfi();
            return None

        }
    };

    let current_high_regs = ...;

    ...

    Some(res)
```

Not a huge diff, but would get rid of one nesting layer.

kaspar030 replied:
yup

elenaf9 commented on src/riot-rs-threads/src/arch/cortex_m.rs:
```diff
@@ -216,53 +212,44 @@ unsafe extern "C" fn PendSV() {
+                }
+            };
+
+            let current_pid = threads.current_pid().unwrap();
```
This will panic once #182 lands. 

As described in #182, it's necessary for the RISC-V implementation that we can differentiate the case when the very first thread starts running. We can only do that if we don't set `current_pid` in the `start_threading` function and instead below once we actually switch to the thread. With that change, `current_pid` would be `None` the very first time this function runs.

Is there any reason why we shouldn't do it that way?

kaspar030 replied:
> Is there any reason why we shouldn't do it that way?

Performance. :)

Let's get measurement in place so any arguments have number. In general, spending even one instruction extra on *every* context switch, to work around needs of the *first* context switch, should be avoided.

Heck, we carried along an idle thread for years in RIOT-c to avoid testing if there's actually a thread. :)

elenaf9 replied:
But `unwrap` internally also simply uses a match to check if it's `Some` or `None`: https://doc.rust-lang.org/src/core/option.rs.html#928.

kaspar030 replied:
fixed in e3966d2864a57a49628b9742e31ac3a9ad8f218c.

elenaf9 commented:
Friendly ping @kaspar030. 

Is anything blocking this PR? I am currently working on an alternative implementation of #241, where these changes are relevant.

Happy to take over the PR and do the rebasing/ solving of merge conflicts.

kaspar030 commented:
> Is anything blocking this PR? I am currently working on an alternative implementation of #241, where these changes are relevant.

I've rebased the PR and updated it to support the `current_thread==None` case. IMO it's good to go.
I did a quick benchmark, on nrf52840dk this uses some cycles less (230 -> 227). With only one thread, some cycles more (80 -> 86). I'd be fine with those numbers as we'll be touching / optimizing this anyways.

elenaf9 reviewed (approved):
LGTM. Thanks!
